___

<a href='https://www.udemy.com/user/Neamatullah-ekhteyari/'><img src='../Neamatullah.png'/></a>
___
<center><em>Content Copyright by Neamatullah Ekhteyari</em></center>

### Python Code for Timing Functions – Detailed Explanation
Python provides several methods for measuring the execution time of your code. Two of the most commonly used methods are ```time``` and ``` timeit```. Below, I will explain the code for timing functions and their output, as well as the purpose of each method.

#### 1. Using time to Measure Execution Time
The time module provides a simple way to measure the elapsed time between two points in your code.

In [36]:
import time

def func1(n):
    return [str(num) for num in range(n)]

def func2(n):
    return list(map(str, range(n)))


start_time = time.time()
func1(10**6)
end_time= time.time()
print(f"funciton 1 time: {end_time - start_time:.5f} seconds")

start_time = time.time()
func2(10**6)
end_time = time.time()
print(f"Function 2 time: {end_time - start_time:.5f} seconds")



funciton 1 time: 0.45623 seconds
Function 2 time: 0.40577 seconds


#### Explanation:

- time.time(): Returns the current time in seconds since the epoch (a floating-point number).

- func1 and func2: Both functions generate a list of string numbers, but they use different methods to do so:

    - func1: Uses a list comprehension.

    - func2: Uses the map() function, which applies the str() function to each element in the range.

#### Timing Logic:

- The start_time is recorded before calling the function, and end_time is recorded after the function finishes.

- The difference end_time - start_time gives the total time taken by the function.

### 2. Using timeit to Measure Code Performance
The ```timeit``` module is designed for more accurate timing, especially for small code snippets. It automatically takes care of the setup and repeated execution to give you more precise results.

In [41]:
import timeit

execution_time = timeit.timeit("sum(range(1000))", number=10000)
print(f"Execution time: {execution_time} seconds")

Execution time: 0.4190204000042286 seconds


#### Explanation:

- ```timeit.timeit()```: Executes the given code snippet a specified number of times (in this case, 10,000).

- ```"sum(range(1000))"```: This is the code being timed. It calculates the sum of numbers from 0 to 999.

- ```number=10000```: This tells ```timeit``` to run the code snippet 10,000 times and then return the total execution time.

This provides the total time taken to run the code 10,000 times, allowing for a more consistent result compared to a single execution.

### 3. Using %%timeit in Jupyter Notebooks
In Jupyter notebooks, ```%%timeit``` is a special "magic" command that automatically times the code in a cell. It runs the code multiple times to get the average execution time.

In [42]:
%%timeit
sum(range(1000))

43.1 μs ± 1.9 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


#### Explanation:

- ```%%timeit```: Measures the execution time of the entire cell.

- The sum(range(1000)) calculates the sum of numbers from 0 to 999.

- The output includes both the average time and the standard deviation across several runs.

This output shows the mean time (43.1 microseconds) along with the standard deviation (1.9 microseconds) after running the code 10,000 times over 7 runs.

### 4. Comparing Two Code Snippets Using timeit
You can compare the execution time of two different approaches using timeit.

In [43]:
import timeit

code1 = "sum(range(1000))"
code2 = "total = 0\nfor i in range (1000): total += i"

time1 = timeit.timeit(code1, number=10000)
time2 = timeit.timeit(code2, number=10000)

print(f"Using sum(rang()): {time1} Sec")
print(f"Using loop: {time2} sec")
      

Using sum(rang()): 0.450485999986995 Sec
Using loop: 1.6508842000039294 sec


#### Explanation:

- ```code1```: Calculates the sum of numbers from 0 to 999 using sum(range(1000)).

- ```code2```: Uses a for loop to sum the numbers manually.

- ```timeit.timeit()```: Runs each code snippet 10,000 times and measures the execution time.


The ```sum(range())``` approach is much faster than using a loop, demonstrating the efficiency of Python’s built-in functions over manual iteration.

### 5. Writing and Closing Files
In the following example, we write to two files using ```open()``` and ```close()```

In [1]:
file1 = open("file1.txt", "w+")
file1.write("this is the first file.")
file1.close()


file2 = open("file2.txt", "w+")
file2.write("this is the second file.")
file2.close()


#### Explanation:

- open("file1.txt", "w+"): Opens the file for reading and writing. If the file doesn't exist, it is created.

- file1.write(): Writes the text into the file.

- file1.close(): Closes the file after writing to ensure the changes are saved.

This code writes text to two files, file1.txt and file2.txt, and then closes them.

#### Key Takeaways:
- time module is simple but effective for measuring time intervals.

- timeit module provides more accurate timing by executing code repeatedly and averaging results.

- %%timeit in Jupyter allows quick timing for an entire cell of code.

- Code Efficiency: You can compare different code snippets using timeit to determine the most efficient approach.